clients/drcachesim/scheduler/noise_generator.cpp

```diff
@@ -0,0 +1,182 @@
+ * ARE DISCLAIMED. IN NO EVENT SHALL VMWARE, INC. OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ */
+
+#include <assert.h>
+#include <climits>
+
+#include "noise_generator.h"
+#include "memref.h"
+#include "trace_entry.h"
+
+namespace dynamorio {
+namespace drmemtrace {
+
+noise_generator_t::noise_generator_t()
+{
+}
+
+noise_generator_t::noise_generator_t(noise_generator_info_t &info)
+    : num_records_to_generate_(info.num_records_to_generate)
+    , pid_(info.pid)
+    , tid_(info.tid)
+{
+}
+
+noise_generator_t::~noise_generator_t()
+{
+}
+
+bool
+noise_generator_t::init()
+{
+    at_eof_ = false;
+    ++*this;
+    return true;
+}
+
+std::string
+noise_generator_t::get_stream_name() const
+{
+    return "noise_generator";
+}
+
+trace_entry_t
+noise_generator_t::generate_trace_entry()
+{
+    // TODO i#7216: this is a temporary trace record that we use as a placeholder until
+    // the logic to generate noise records is in place.
+    trace_entry_t generated_entry = { TRACE_TYPE_READ, 4, { 0xdeadbeef } };
+    return generated_entry;
+}
+
+trace_entry_t *
+noise_generator_t::read_next_entry()
+{
+    if (num_records_to_generate_ == 0) {
+        at_eof_ = true;
+        return nullptr;
+    }
+
+    // Do not change the order for generating TRACE_TYPE_THREAD and TRACE_TYPE_PID.
+    // The scheduler expects a tid first and then a pid.
+    if (!tid_generated_) {
+        entry_ = { TRACE_TYPE_THREAD, sizeof(int), { tid_ } };
+        tid_generated_ = true;
+        return &entry_;
+    }
+    if (!pid_generated_) {
+        entry_ = { TRACE_TYPE_PID, sizeof(int), { pid_ } };
+        pid_generated_ = true;
+        return &entry_;
+    }
+    // The scheduler expects a TRACE_MARKER_TYPE_TIMESTAMP for relative threads order.
+    // We provide one with a high value to indicate that noise generator threads have
+    // no dependencies with other threads. The scheduler will re-write these values.
+    if (!marker_timestamp_generated_) {
+        entry_ = { TRACE_TYPE_MARKER,
+                   TRACE_MARKER_TYPE_TIMESTAMP,
+                   { static_cast<addr_t>(ULONG_MAX - 1) } };
+        marker_timestamp_generated_ = true;
+        return &entry_;
+    }
+
+    entry_ = generate_trace_entry();
+
+    if (num_records_to_generate_ == 1) {
+        entry_ = { TRACE_TYPE_THREAD_EXIT, sizeof(int), { tid_ } };
+    }
+    --num_records_to_generate_;
+
+    return &entry_;
+}
+
+template <typename RecordType, typename ReaderType>
```

Comment:
Add a banner comment line to separate the factory from the other class (use the banner style in the style guide; see examples in scheduler_impl.cpp `Specialization for ...`, etc.)